core/src/banking_stage/transaction_scheduler/transaction_state.rs

```diff
@@ -0,0 +1,307 @@
+use {
+    solana_runtime::transaction_priority_details::TransactionPriorityDetails,
+    solana_sdk::{slot_history::Slot, transaction::SanitizedTransaction},
+};
+
+/// Simple wrapper type to tie a sanitized transaction to max age slot.
+pub(crate) struct SanitizedTransactionTTL {
+    pub(crate) transaction: SanitizedTransaction,
+    pub(crate) max_age_slot: Slot,
+}
+
+/// TransactionState is used to track the state of a transaction in the transaction scheduler
+/// and banking stage as a whole.
+///
+/// There are two states a transaction can be in:
+///     1. `Unprocessed` - The transaction is available for scheduling.
+///     2. `Pending` - The transaction is currently scheduled or being processed.
+///
+/// Newly received transactions are initially in the `Unprocessed` state.
+/// When a transaction is scheduled, it is transitioned to the `Pending` state,
+///   using the `transition_to_pending` method.
+/// When a transaction finishes processing it may be retryable. If it is retryable,
+///   the transaction is transitioned back to the `Unprocessed` state using the
+///   `transition_to_unprocessed` method. If it is not retryable, the state should
+///   be dropped.
+///
+/// For performance, when a transaction is transitioned to the `Pending` state, the
+///   internal `SanitizedTransaction` is moved out of the `TransactionState` and sent
+///   to the appropriate thread for processing. This is done to avoid cloning the
+///  `SanitizedTransaction`.
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum TransactionState {
+    /// The transaction is available for scheduling.
+    Unprocessed {
+        transaction_ttl: SanitizedTransactionTTL,
+        transaction_priority_details: TransactionPriorityDetails,
+        forwarded: bool,
+    },
+    /// The transaction is currently scheduled or being processed.
+    Pending {
+        transaction_priority_details: TransactionPriorityDetails,
```

Comment:
Naming was hard here, I'm not sure `Processing` is accurate either.

All this means is that we've selected the transaction:
1. It may still be in the scheduler thread, not sent via channel as we are building the batch
2. It may be queue in the channel, but not actually processing yet because there's other work being done by that thread
3. It is being processed by that thread
4. It's been processed, but scheduler has not yet received the finished work message

I actually used the name "InFlight" in a separate component which was tracking some batched transaction information.
Do you think `InFlight` would be a better name here?

Reply:
`InFlight` captures 1 & 2 better, but not so much for 3 & 4. I'm good for leaving it as `pending` for now.